crates/nu-parser/src/parser.rs

```diff
@@ -5272,15 +5272,43 @@ pub fn parse_record(working_set: &mut StateWorkingSet, span: Span) -> Expression
 
             idx += 1;
             if idx == tokens.len() {
-                working_set.error(ParseError::Expected("record", span));
-                return garbage(span);
+                working_set.error(ParseError::Expected(
+                    "':'",
+                    Span::new(curr_span.end, curr_span.end),
+                ));
+                output.push(RecordItem::Pair(
+                    garbage(curr_span),
+                    garbage(Span::new(curr_span.end, curr_span.end)),
+                ));
+                break;
             }
-            let colon = working_set.get_span_contents(tokens[idx].span);
+            let colon_span = tokens[idx].span;
+            let colon = working_set.get_span_contents(colon_span);
             idx += 1;
-            if idx == tokens.len() || colon != b":" {
-                //FIXME: need better error
-                working_set.error(ParseError::Expected("record", span));
-                return garbage(span);
+            if colon != b":" {
+                working_set.error(ParseError::Expected(
+                    "':'",
+                    Span::new(colon_span.start, colon_span.start),
+                ));
+                output.push(RecordItem::Pair(
+                    field,
+                    garbage(Span::new(
+                        colon_span.start,
+                        tokens[tokens.len() - 1].span.end,
+                    )),
+                ));
+                break;
+            }
+            if idx == tokens.len() {
+                working_set.error(ParseError::Expected(
+                    "value for record field",
+                    Span::new(colon_span.end, colon_span.end),
+                ));
+                output.push(RecordItem::Pair(
+                    garbage(Span::new(curr_span.start, colon_span.end)),
+                    garbage(Span::new(colon_span.end, tokens[tokens.len() - 1].span.end)),
+                ));
+                break;
```

Comment:
Just to check, does this possibly affect downstream code that this keeps `output` semi intact versus in the old code completely aborting with `Expression::Garbage`?

Reply:
I don't think so. The only other place where garbage isn't returned but merely added to an otherwise valid-looking expression is [these lines](https://github.com/nushell/nushell/blob/main/crates/nu-parser/src/parser.rs#L4838-L4841) in `parse_math_expression`. However, there are definitely other situations in which the record end up semi-intact while containing an `Expression::Garbage`, such as `{ x: $non_existent_var }`, and that works.

Also, it looks like the [repl code](https://github.com/nushell/nushell/blob/main/crates/nu-cli/src/repl.rs#L547) checks for parse errors before evaluating. There's a couple other places that don't check for parse errors (syntax_highlight.rs and completer.rs) but they don't check if the returned expression is garbage either, so it's probably fine 🤷

Reply:
OK sounds good to me.

Thanks for clarifying.